web/recv.js

```diff
@@ -0,0 +1,401 @@
+var Sink = function () {
+
+  var _fountainBuff = undefined;
+  var _errBuff = undefined;
+
+  // public interface
+  return {
+    on_decode: function (buff) {
+      if (_fountainBuff === undefined) {
+        const size = Module._cimbard_get_bufsize(); // max length of buff. We could also resize as we go...
+        const dataPtr = Module._malloc(size);
+        _fountainBuff = new Uint8Array(Module.HEAPU8.buffer, dataPtr, size);
+      }
+      if (buff.length == 0) { // sanity check
+        return;
+      }
+      _fountainBuff.set(buff);
+
+      console.log('sink decode ' + _fountainBuff); //TODO: base64?
+      var res = Module._cimbard_fountain_decode(_fountainBuff.byteOffset, buff.length);
+      console.log("on decode got res " + res);
+
+      const report = Sink.get_report();
+      if (Array.isArray(report)) {
+        Recv.render_progress(report);
+      }
+      else {
+        Recv.set_HTML("tdec", "decode " + res + ". " + report);
+      }
+
+      if (res > 0) {
+        const res32t = Number(res & 0xFFFFFFFFn);; // truncate BigInt res (int64_t) to a uint32_t
+        Sink.reassemble_file(res32t);
+      }
+    },
+
+    get_report: function () {
+      const maxSize = 1024;
+      if (_errBuff === undefined) {
+        _errBuff = Module._malloc(maxSize);
+      }
+      const errlen = Module._cimbard_get_report(_errBuff, maxSize);
+      if (errlen > 0) {
+        const errview = new Uint8Array(Module.HEAPU8.buffer, _errBuff, errlen);
+        const td = new TextDecoder();
+        const text = td.decode(errview);
+        try {
+          return JSON.parse(text);
+        } catch (error) {
+          return text;
+        }
+      }
+    },
+
+    reassemble_file: function (id) {
+      const size = Module._cimbard_get_filesize(id);
+      //alert("we did it!?! " + size);
+      const dataPtr = Module._malloc(size);
+      const buff = new Uint8Array(Module.HEAPU8.buffer, dataPtr, size);
+      try {
+        var res = Module._cimbard_finish_copy(id, buff.byteOffset, buff.length);
+        if (res < 0) {
+          alert("reassemble_file failed :(" + res);
+          console.log("we biffed it. :( " + res);
+          Recv.set_HTML("errorbox", "reassemble_file failed :( " + res);
+        }
+        else {
+          //Recv.download_bytes(buff, size + ".zst"); // size -> name, eventually
+          Zstd.decompress(id + "." + size, buff);
+        }
+      } catch (error) {
+        console.log("failed finish copy or download?? " + error);
+      }
+      // this needs to happen after decompress() completes
+      // currently decompress is sync, so it's fine. But...
+      Module._free(dataPtr);
```

Comment:
^

This is a bit of a diabolical (potential) race condition. It's easy to imagine this coming back to bite us in some future change, but it's hard to see what the right fix is until we hit the (future) problem.